### ISSUE TITLE
Language setting not applied when signing in via SSO

### ISSUE THREAD

**Issue by salfers** [netbox-community/netbox#16995]
### Deployment Type

Self-hosted

### NetBox Version

v4.0.7

### Python Version

3.11

### Steps to Reproduce

1. Set your user language to something that differs from the automatically detected language
2. Clear cookies
3. Log into Netbox via **Google** SSO

Bonus: if you log in via password everything works correctly.

### Expected Behavior

4\. Language from user settings is applied

### Observed Behavior

4\. The automatically detected language is used

You can see that the language cookie is not being created:
![grafik](https://github.com/user-attachments/assets/d34fde6a-7ce4-40c4-a420-da88532d0ff8)

**Comment by benhur1999** [comment]
I can confirm this behavior with external auth. 

For reference, we have the following set in the netbox config:
```
REMOTE_AUTH_ENABLED = True
REMOTE_AUTH_BACKEND = 'netbox.authentication.RemoteUserBackend'
REMOTE_AUTH_HEADER = 'HTTP_X_REMOTE_USER'

```

Environment: self-hosted, Netbox 4.0.7, Python 3.12

**Comment by m2martin** [comment]
@salfers I checked it using AzureAD and cannot confirm/reproduce your behavior.

Was your user created by SSO or did it already exist?
Can you please also post your configuration parameters starting with `REMOTE_AUTH_` (omit any sensitive data, of course)?

**Comment by m2martin** [comment]
With `REMOTE_AUTH_BACKEND = 'netbox.authentication.RemoteUserBackend'` set, I can partially reproduce it. I can see the first page being loaded with the browser's language but any subsequent request will be displayed in the language from the user's properties.
Can you confirm this behavior?

**Comment by benhur1999** [comment]
@m2martin No, I cannot confirm the behavior. The language stays on the browser default and the user preferences have no impact.

What I did
 1, open new Incognito-Tab in Chrome
 2. open Netbox URL and authenticate using HTTP Auth
 3. click on Devices ("Geräte")menu entry  in the Devices ("Geräte") section
 4. click on the Netbox logo menu 
 
No change in language. I can only get it to switch to English (the users perference) by opening the user preferences and then Clicking "save"

**Comment by m2martin** [comment]
I'm doing the same and can not confirm it. Can you please try it with v4.0.8? This is the most obvious difference between our tests.
(Maybe fixed with #16732)

**Comment by benhur1999** [comment]
@m2martin I'm running 4.0.8 since a few days. There was no change in behavior  between 4.0.8 and 4.0.7.

**Comment by leo-pempera** [comment]
We are having the same issue.
Browser language is German, but netbox DEFAULT_LANGUAGE is set to en-us.
Upon login with SSO netbox is in german, not only the first site after login.
In the user settings the language is set to english. Upon saving them without changing anything the correct language is applied.
This resets every few days and the user has to change apply the user settings again.

**Comment by benhur1999** [comment]
FWIW, I also have set `DEFAULT_LANGUAGE = 'en-us'`

**Comment by m2martin** [comment]
I think, I found it. Is `LOGIN_PERSISTENCE = False`?

**Comment by benhur1999** [comment]
@m2martin Yes, I have `LOGIN_PERSISTENCE = False`

Other settings:
```
LOGIN_REQUIRED = True
LOGIN_TIMEOUT = None
```

**Comment by m2martin** [comment]
@arthanson I have a fix ready. You can assign it to me.

Root cause:
Language cookies will only be set on regular logins by the login view or renewed when having `LOGIN_PERSISTENCE=True`.
Setting the cookie in every workflow for every possible authentication method (local, OAuth, LDAP, etc.) is hard to maintain.
With the fix the language handling was moved to the `CoreMiddleware` to have it processed independent from the authentication method/backend.

**Comment by jeremystretch** [comment]
@m2martin I've assigned this to you, thanks.

**Comment by ISNL-PimP** [comment]
Can this issue thread be revived? The issue is still present.
When I sign in using SSO, my NetBox display language is reset to my browser display language. If I open NetBox' settings, the language setting is still displayed as 'English' while the UI is in Dutch.

![Image](https://github.com/user-attachments/assets/df4be336-69bb-4828-9f1d-45ab93c3d2bc)

**Comment by jeremystretch** [comment]
@ISNL-PimP this issue was closed for inactivity as the initial PR was not accepted. Would you like to volunteer to own this bug and submit a fix?

**Comment by m2martin** [comment]
@jeremystretch There is a closed PR [#16995](https://github.com/netbox-community/netbox/pull/17122) - shall we go that way? If yes, I may do a new attempt for the current version.
It has been too long ago, so I'll have to do it from scratch - at least to understand the problem. But I can remember that the suggested change was not accepted.

**Comment by jeremystretch** [comment]
I'm not in a position to offer any direction unfortunately. But this issue does need a volunteer committed to delivering a fix before it can be reopened.

**Comment by m2martin** [comment]
Hi @jeremystretch,

I've done a lot of tests today and was able to reproduce the behavior, even in the current version.

There are corner cases depending on which authentication method is used. My old solution which moves the language activation to the middleware (in a slightly varying way) does solve the issue.

Tested it with:
- Local auth
- Remote auth (OAuth2, EntraID)
- Remote auth (HTTP header auth)

I'd volunteer to present my new version of the fix in a new PR.

**Comment by m2martin** [comment]
Due to the contribution guidelines I did not open a PR without being assigned to an open issue.
Can we please re-open the issue to bring the fix forward?

**Comment by vink78** [comment]
I can confirm your issue is still present on v4.4.1.

A rebased version of https://github.com/m2martin/netbox/commit/526fb7c95325091abf62be18a3f82c9584886d9e fixes the issue.

**Comment by joeladria** [comment]
possible to reopen this issue since it sounds like we have a working fix that just needs the issue assigned to it?

**Comment by adepretis** [comment]
Still a bug - somebody needs to reopen this issue. Seems to be a quick thing with fixes already available.

**Comment by jeremystretch** [comment]
@adepretis I'm happy to assign this to you if you're volunteering to submit a PR.

**Comment by adepretis** [comment]
@jeremystretch As I see it @m2martin already asked to be assigned to that issue so he can submit an MR for the fix he already has (see https://github.com/netbox-community/netbox/issues/16995#issuecomment-3026551532)

**Comment by jeremystretch** [comment]
We've already assigned it to him, and his PR was not accepted. My question for you @adepretis, as you are the only person to comment on this in recent months, is whether _you_ are volunteering to own this.

**Comment by adepretis** [comment]
Ah, sorry, I didn't dig that deep and didn't see the MR ;-)

Unfortunately I can not accept the offer, as I am not proficient with Django and the issue that came up with the past MR by @m2martin seems rather complicated.

So, I guess, this will stay unresolved then. Maybe I'll give Claude a try.

**Comment by m2martin** [comment]
Incorrect - since I asked for getting assigned in July 2025 I was not.

Edit: As everyone can see my proposed change, I'd rather save my time for other projects than spending time to align the change with the current code base just to have it declined.
I'm not going to reinvent the wheel.

**Comment by jeremystretch** [comment]
@m2martin you were not reassigned because your first PR was not accepted, and you [indicated](https://github.com/netbox-community/netbox/issues/16995#issuecomment-2974182668) that you were not clear on how to proceed.

**Comment by m2martin** [comment]
@jeremystretch Short wrap-up:
The PR which was not accepted was #17122.

I remember that I spent a lot of time afterwards - almost a year later - with testing the different authentication backends while keeping the necessary change as small as possible after the issue came up again here.

AFAIK, this is the most recent version (didn't touch it since then) https://github.com/m2martin/netbox/commit/526fb7c95325091abf62be18a3f82c9584886d9e

I was not assigned to the issue after asking for it and no PR was created.

If you think its worth it, I'd again check if the change is applicable to the current version and create another MR.
But again, this would be the third attempt and the third time I have to dig into it. If it's not likely to get accepted, I'd rather spend my time on other things.

**Comment by vink78** [comment]
> I can confirm your issue is still present on v4.4.1.
> 
> A rebased version of [m2martin@526fb7c](https://github.com/m2martin/netbox/commit/526fb7c95325091abf62be18a3f82c9584886d9e) fixes the issue.

Since #21590 It seems that the fix is no more needed (tested on Netbox 4.5.8).

Other can confirm the issue is fixed ?
